File: repertoire/data-integration/providers/quality_rule/cross_field.ts
// Quality Rule Provider: Cross-Field Validation
// Evaluates multi-field validation expressions across record fields.
// Dimension: consistency

export const PROVIDER_ID = 'cross_field';
export const PLUGIN_TYPE = 'quality_rule';

export interface FieldDef {
  name: string;
  type: string;
  required?: boolean;
  constraints?: Record<string, unknown>;
}

export interface RuleConfig {
  options?: Record<string, unknown>;
  threshold?: number;
}

export interface RuleResult {
  valid: boolean;
  message?: string;
  severity: 'error' | 'warning' | 'info';
}

export type QualityDimension = 'completeness' | 'uniqueness' | 'validity' | 'consistency' | 'timeliness' | 'accuracy';

export class CrossFieldQualityProvider {
  validate(
    _value: unknown,
    field: FieldDef,
    record: Record<string, unknown>,
    config: RuleConfig
  ): RuleResult {
    const expression = config.options?.expression as string | undefined;
    const fields = config.options?.fields as string[] | undefined;

    if (!expression) {
      return {
        valid: false,
        message: `Cross-field rule for '${field.name}' is misconfigured: no expression provided.`,
        severity: 'error',
      };
    }

    if (!fields || fields.length === 0) {
      return {
        valid: false,
        message: `Cross-field rule for '${field.name}' is misconfigured: no fields specified.`,
        severity: 'error',
      };
    }

    try {
      const result = this.evaluateExpression(expression, fields, record);
      if (!result) {
        return {
          valid: false,
          message: `Cross-field validation failed for '${field.name}': expression '${expression}' evaluated to false.`,
          severity: 'error',
        };
      }
      return { valid: true, severity: 'error' };
    } catch (err) {
      return {
        valid: false,
        message: `Cross-field rule evaluation error for '${field.name}': ${(err as Error).message}`,
        severity: 'error',
      };
    }
  }

  private evaluateExpression(
    expression: string,
    fields: string[],
    record: Record<string, unknown>
  ): boolean {
    // Handle comparison expressions: field_a > field_b, field_a >= field_b, etc.
    const comparisonMatch = expression.match(
      /^(\w+)\s*(>|>=|<|<=|==|!=)\s*(\w+)$/
    );
    if (comparisonMatch) {
      const [, leftField, operator, rightField] = comparisonMatch;
      const leftVal = this.resolveValue(record[leftField]);
      const rightVal = this.resolveValue(record[rightField]);

      if (leftVal === null || rightVal === null) return true;

      switch (operator) {
        case '>': return leftVal > rightVal;
        case '>=': return leftVal >= rightVal;
        case '<': return leftVal < rightVal;
        case '<=': return leftVal <= rightVal;
        case '==': return leftVal === rightVal;
        case '!=': return leftVal !== rightVal;
        default: return false;
      }
    }

    // Handle conditional required: if field_a == "value" then field_b required
    const conditionalMatch = expression.match(
      /^if\s+(\w+)\s*==\s*"([^"]*)"\s+then\s+(\w+)\s+required$/
    );
    if (conditionalMatch) {
      const [, condField, condValue, reqField] = conditionalMatch;
      const fieldVal = record[condField];
      if (String(fieldVal) === condValue) {
        const reqVal = record[reqField];
        return reqVal !== null && reqVal !== undefined && reqVal !== '';
      }
      return true;
    }

    // Handle mutual exclusion: exactly_one_of(field_a, field_b, ...)
    const mutexMatch = expression.match(/^exactly_one_of\(([^)]+)\)$/);
    if (mutexMatch) {
      const mutexFields = mutexMatch[1].split(',').map((f) => f.trim());
      const presentCount = mutexFields.filter((f) => {
        const v = record[f];
        return v !== null && v !== undefined && v !== '';
      }).length;
      return presentCount === 1;
    }

    // Handle at_least_one_of(field_a, field_b, ...)
    const atLeastMatch = expression.match(/^at_least_one_of\(([^)]+)\)$/);
    if (atLeastMatch) {
      const checkFields = atLeastMatch[1].split(',').map((f) => f.trim());
      return checkFields.some((f) => {
        const v = record[f];
        return v !== null && v !== undefined && v !== '';
      });
    }

    // Fallback: use Function constructor for simple JS boolean expressions
    const context: Record<string, unknown> = {};
    for (const f of fields) {
      context[f] = record[f];
    }
    const paramNames = Object.keys(context);
    const paramValues = Object.values(context);
    const fn = new Function(...paramNames, `return Boolean(${expression});`);
    return fn(...paramValues);
  }

  private resolveValue(val: unknown): number | null {
    if (val === null || val === undefined) return null;
    if (typeof val === 'number') return val;
    const dateTs = Date.parse(String(val));
    if (!isNaN(dateTs)) return dateTs;
    const num = Number(val);
    return isNaN(num) ? null : num;
  }

  appliesTo(_field: FieldDef): boolean {
    return true;
  }

  dimension(): QualityDimension {
    return 'consistency';
  }
}

export default CrossFieldQualityProvider;

File: repertoire/concepts/data-integration/providers/quality_rule/cross_field.rs
// Quality Rule Provider: Cross-Field Validation
// Evaluates multi-field validation expressions across record fields.
// Dimension: consistency

use std::collections::HashMap;

pub const PROVIDER_ID: &str = "cross_field";
pub const PLUGIN_TYPE: &str = "quality_rule";

#[derive(Debug, Clone)]
pub struct FieldDef {
    pub name: String,
    pub field_type: String,
    pub required: Option<bool>,
    pub constraints: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct RuleConfig {
    pub options: Option<HashMap<String, serde_json::Value>>,
    pub threshold: Option<f64>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Severity { Error, Warning, Info }

#[derive(Debug, Clone)]
pub struct RuleResult {
    pub valid: bool,
    pub message: Option<String>,
    pub severity: Severity,
}

#[derive(Debug, Clone, PartialEq)]
pub enum QualityDimension {
    Completeness, Uniqueness, Validity, Consistency, Timeliness, Accuracy,
}

pub struct CrossFieldQualityProvider;

impl CrossFieldQualityProvider {
    pub fn new() -> Self {
        Self
    }

    pub fn validate(
        &self,
        _value: &serde_json::Value,
        field: &FieldDef,
        record: &HashMap<String, serde_json::Value>,
        config: &RuleConfig,
    ) -> RuleResult {
        let expression = match config.options.as_ref()
            .and_then(|o| o.get("expression"))
            .and_then(|v| v.as_str())
        {
            Some(e) => e.to_string(),
            None => return RuleResult {
                valid: false,
                message: Some(format!(
                    "Cross-field rule for '{}' is misconfigured: no expression provided.",
                    field.name
                )),
                severity: Severity::Error,
            },
        };

        let fields: Vec<String> = match config.options.as_ref()
            .and_then(|o| o.get("fields"))
            .and_then(|v| v.as_array())
        {
            Some(arr) => arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect(),
            None => return RuleResult {
                valid: false,
                message: Some(format!(
                    "Cross-field rule for '{}' is misconfigured: no fields specified.",
                    field.name
                )),
                severity: Severity::Error,
            },
        };

        if fields.is_empty() {
            return RuleResult {
                valid: false,
                message: Some(format!(
                    "Cross-field rule for '{}' is misconfigured: no fields specified.",
                    field.name
                )),
                severity: Severity::Error,
            };
        }

        match self.evaluate_expression(&expression, &fields, record) {
            Ok(true) => RuleResult { valid: true, message: None, severity: Severity::Error },
            Ok(false) => RuleResult {
                valid: false,
                message: Some(format!(
                    "Cross-field validation failed for '{}': expression '{}' evaluated to false.",
                    field.name, expression
                )),
                severity: Severity::Error,
            },
            Err(err) => RuleResult {
                valid: false,
                message: Some(format!(
                    "Cross-field rule evaluation error for '{}': {}",
                    field.name, err
                )),
                severity: Severity::Error,
            },
        }
    }

    fn evaluate_expression(
        &self,
        expression: &str,
        _fields: &[String],
        record: &HashMap<String, serde_json::Value>,
    ) -> Result<bool, String> {
        // Handle comparison: field_a > field_b, field_a >= field_b, etc.
        let comparison_ops = [">=", "<=", "!=", "==", ">", "<"];
        for op in &comparison_ops {
            if let Some(pos) = expression.find(op) {
                let left_field = expression[..pos].trim();
                let right_field = expression[pos + op.len()..].trim();

                // Check if this is a conditional expression (starts with "if")
                if left_field.starts_with("if ") {
                    continue;
                }

                let left_val = self.resolve_value(record.get(left_field));
                let right_val = self.resolve_value(record.get(right_field));

                match (left_val, right_val) {
                    (Some(l), Some(r)) => {
                        let result = match *op {
                            ">" => l > r,
                            ">=" => l >= r,
                            "<" => l < r,
                            "<=" => l <= r,
                            "==" => (l - r).abs() < f64::EPSILON,
                            "!=" => (l - r).abs() >= f64::EPSILON,
                            _ => return Err(format!("Unknown operator: {}", op)),
                        };
                        return Ok(result);
                    }
                    _ => return Ok(true), // Skip if fields are null
                }
            }
        }

        // Handle: if field_a == "value" then field_b required
        if expression.starts_with("if ") && expression.contains(" then ") {
            let parts: Vec<&str> = expression.splitn(2, " then ").collect();
            if parts.len() == 2 {
                let condition = parts[0].trim_start_matches("if ").trim();
                let action = parts[1].trim();

                // Parse condition: field == "value"
                if let Some(eq_pos) = condition.find("==") {
                    let cond_field = condition[..eq_pos].trim();
                    let cond_value = condition[eq_pos + 2..].trim()
                        .trim_matches('"')
                        .trim_matches('\'');

                    let field_val = record.get(cond_field)
                        .and_then(|v| v.as_str())
                        .unwrap_or("");

                    if field_val == cond_value {
                        // Parse action: field_b required
                        if action.ends_with(" required") {
                            let req_field = action.trim_end_matches(" required").trim();
                            let req_val = record.get(req_field);
                            let is_present = match req_val {
                                None => false,
                                Some(v) => !v.is_null() && v.as_str().map_or(true, |s| !s.is_empty()),
                            };
                            return Ok(is_present);
                        }
                    } else {
                        return Ok(true); // Condition not met, rule passes
                    }
                }
            }
        }

        Err(format!("Unsupported expression format: {}", expression))
    }

    fn resolve_value(&self, val: Option<&serde_json::Value>) -> Option<f64> {
        match val? {
            serde_json::Value::Number(n) => n.as_f64(),
            serde_json::Value::String(s) => {
                // Try as number first, then as date string for comparison
                s.parse::<f64>().ok()
            }
            _ => None,
        }
    }

    pub fn applies_to(&self, _field: &FieldDef) -> bool {
        true
    }

    pub fn dimension(&self) -> QualityDimension {
        QualityDimension::Consistency
    }
}

impl Default for CrossFieldQualityProvider {
    fn default() -> Self {
        Self::new()
    }
}
